### ISSUE TITLE
Actor starvation when self-messages and different dispatchers are involved

### ISSUE THREAD

**Issue by clockfly** [akka/akka#17341]
## Problem Statement:

We have a benchmark test which process 10 million message per second(project https://github.com/intel-hadoop/gearpump), during the test, I found sometimes some message cannot be delivered to target actor.

## Test case description:
The test is like this:
I have 600 upstream actor, and 600 downstream actors distribued in 60 JVM. the upstream actors will crazily send data to downstream actors in a shuffle way. The upstream actors rely on self-sending message to trigger the repeation of sending. 

upstream tasks and downstream tasks will share a custom dispatcher(not default dispatcher, or remote-default-dispatcher).

Upstream Actor logic:
1. upstream actor A send itself a tick message
2. in receive of A, sending one messge to downstream actors. After sending, send itself another tick message
3. repeat step 2

## Trouble-shooting and findings
I compiled akka source to add log trace, and confirmed that the lost message is already inserted into target Mailbox queue(SingleConsumerOnlyUnboundedMailbox), and the Malibox(Runnable) is already inserted into internal task queue of fork-join pool. **The strange part** is that the mbox was **NOT** get scheduled to a real thead and run.

### Finding1: about akka Version
I find version akka 2.3.6 don't have this problem, while version 2.3.9 has.

### Finding2: about self-sending message
For my test, the upstream actors need to send themselves repeatedly tick message so that they can trigger themselves to send more message downstream. If I change the self-sending message, to be an indirect self sending, then the problem is gone. 
For example, instead of a sending directly to itself A->A, I create a intermediate actor C, A->C->A,

### Finding 3: About changes in latest akka version
Since 2.3.6 doesn't have this problem, I browser through the akka history, and find a suspicious change. https://github.com/akka/akka/pull/16152/files#diff-2c9d0638ce3a1f007623ac9d1ead1725R379

The intent of this PR is to make MailBox a subclass of ForkjoinTask, so that we can optimize the performance by fully using fork-join ideas. I doubt the changes in this PR will also create fairness problem. 

Here is some change in AkkaForkJoinPool
```scala
     extends ForkJoinPool(parallelism, threadFactory, unhandledExceptionHandler, true) with LoadMetrics {
-    override def execute(r: Runnable): Unit =		+    override def execute(r: Runnable): Unit = {
-      if (r eq null) throw new NullPointerException else super.execute(new AkkaForkJoinTask(r))		+      if (r eq null) throw new NullPointerException("The Runnable must not be null")
+      val task =
+        if (r.isInstanceOf[ForkJoinTask[_]]) r.asInstanceOf[ForkJoinTask[Any]]
+        else new AkkaForkJoinTask(r)
+      Thread.currentThread match {
+        case worker: ForkJoinWorkerThread if worker.getPool eq this ⇒ task.fork()
+        case _ ⇒ super.execute(task)
+      }
```

If actor A is sending itself message, then it will trigger AkkaForkJoinPool.execute, as the calling thread is already in the fork-join pool, 
```scala
+      Thread.currentThread match {
+        case worker: ForkJoinWorkerThread if worker.getPool eq this ⇒ task.fork()
+        case _ ⇒ super.execute(task)
+      }
```

So it will pick first branch which execute ```task.fork()```

I doubt task.fork() will have higher priority than the other branch "super.execute(task)", so possibly the later will not get scheduled.

Thanks in advance.
Sean

**Comment by MQ-EL** [comment]
I think when Actor processing received message in forkjoin pool, nobody include itself can tell msg to it process in parallel. 
 so, actor A ! msg not trigger AkkaForkJoinPool.execute, it just enqueue msg into mailbox, and just return. 
you can see registerForExecution method in Dispatcher class.

you can switch config from forkjoin to normal pool in 2.3.9, then test. So, exclude other suspect firstly.

Just some of my little thinks

**Comment by clockfly** [comment]
@MQ-EL 

Dispatcher.registerForExecution will try to execute mbox with executorservice.
Check:
```scala
protected[akka] override def registerForExecution(mbox: Mailbox, hasMessageHint: Boolean, hasSystemMessageHint: Boolean): Boolean = {
    if (mbox.canBeScheduledForExecution(hasMessageHint, hasSystemMessageHint)) { //This needs to be here to ensure thread safety and no races
      if (mbox.setAsScheduled()) {
        try {
          executorService execute mbox
          true
        } catch {
          case e: RejectedExecutionException ⇒
            try {
              executorService execute mbox
              true
            } catch { //Retry once
              case e: RejectedExecutionException ⇒
                mbox.setAsIdle()
                eventStream.publish(Error(e, getClass.getName, getClass, "registerForExecution was rejected twice!"))
                throw e
            }
        }
      } else false
    } else false
  }
```

**Comment by MQ-EL** [comment]
when A tell msg to itself, A's mbox.setAsScheduled() will return false. because mailbox has set currentStatus flag with Scheduled value

**Comment by MQ-EL** [comment]
eh.... A ! msg to itself in scheduler's thread in your test?

**Comment by clockfly** [comment]
@MQ-EL 

In Mailbox.run
```scala

  final def run = {
    try {
      if (!isClosed) { //Volatile read, needed here
        processAllSystemMessages() //First, deal with any system messages
        processMailbox() //Then deal with messages
      }
    } finally {
      setAsIdle() //Volatile write, needed here
      dispatcher.registerForExecution(this, false, false)
    }
  }
```

in finally, it will set the status as idle, and then do ```dispatcher.registerForExecution(this, false, false),```
This call will schedule the mbox to Executor Service again.

**Comment by MQ-EL** [comment]
Yes, you are right. but it's not the problem. it's sequence!!! The msg will be exec eventually, that's we wanted. 

I want to know which msg is lost? tick or A->B msg?

**Comment by clockfly** [comment]
@MQ-EL 

```
I want to know which msg is lost? tick or A->B msg?
```
I see your questions now.  

Suppose there is another dispatcher(for example: from other remote actors) sending a ad-hoc message to A, that message may not be delivered to A.

**Comment by drewhk** [comment]
Hi Sean,

Is there a short reproducer we can try? I would like to see the whole
scenario in detail, preferably in code.

-Endre

**Comment by clockfly** [comment]
@drewhk 

I will try to make one.

**Comment by clockfly** [comment]
Here is the reproducer I just wrote:

```scala
package test 

import akka.actor.{ActorSystem, Props, Actor}
import com.typesafe.config.ConfigFactory

object ActorHang {

  val config = ConfigFactory.parseString(
"""
  |actorhang {
  |
  |  task-dispatcher {
  |    mailbox-type = "akka.dispatch.SingleConsumerOnlyUnboundedMailbox"
  |    throughput = 100
  |    fork-join-executor {
  |      parallelism-factor = 2
  |      parallelism-max = 2
  |      parallelism-min = 2
  |    }
  |  }
  |}
""".stripMargin
  )

  def main(args: Array[String]): Unit = {
    val system = ActorSystem("hang", config)

    val tasks = (0 until 20).map { i =>
      system.actorOf(Props(new Task(i)).withDispatcher("actorhang.task-dispatcher"))
    }.toList

    val messageLossTask = system.actorOf(Props(new MessageLossTask).withDispatcher("actorhang.task-dispatcher"))

    //wait until all task actor is running
    Thread.sleep(3000)

    Console.println("Sending CheckMessageLoss to target task")
    messageLossTask  ! CheckMessageLoss

    system.awaitTermination()
  }

  class Task(id: Int) extends Actor {
    private var sent = 0

    self ! "tick"

    override def receive: Receive = {
      case "tick" =>
        sent += 1
        self ! "tick"
    }
  }

  class MessageLossTask extends Actor {

    override def receive: Receive = {
      case msg: Any =>
        // In 2.3.9, CheckMessageLoss will not be received
        Console.println(s"Receiving message $msg")
    }
  }

  case object CheckMessageLoss
}
```

**Comment by drewhk** [comment]
Thank you, we will investigate it

**Comment by MQ-EL** [comment]
In FJPoll (scala 2.3.9), the size of task queues is larger than the number of cores, the argorithm as follows:

```java
int p = the number of cores;         // find power of two table size
int n = (p > 1) ? p - 1 : 1;    // n represent the queue size 
n |= n >>> 1; n |= n >>> 2;  n |= n >>> 4;
n |= n >>> 8; n |= n >>> 16; n = (n + 1) << 1;
```

When a FJThread created, it's new workqueue is created and register in FJPoll queues ( has collision detection).

The result is the FJPool queues has two type queues, one for outsider submission, and the other is owned by it's FJThread. when subtask is forked, it also pushed into FJThread's queues directly!
When FJThread start, it handles all task from it's own queue in one loop: 
```java
final void runTask(ForkJoinTask<?> t) {
            if (t != null) {
                (currentSteal = t).doExec();
                currentSteal = null;
                ++nsteals;
                if (base - top < 0) {       // process remaining local tasks
                    if (mode == 0)
                        popAndExecAll();
                    else
                        pollAndExecAll();
                }
            }
        }
```
if forked task in queue was produced endlessly, the FJThread will never go to next loops. So, it has not chance to steal outsider submission tasks(starvation happens):
```java
final void runWorker(WorkQueue w) {
        w.growArray(); // allocate queue
        do { w.runTask(scan(w)); } while (w.qlock >= 0);
    }
```

But think FJP name, does Dog lea design it intentionly, the original fork task was first class task than  external task? 
```scala
package com.sample.test

import scala.concurrent.forkjoin.{ForkJoinPool, ForkJoinTask}

object ForkjoinTester extends App {

  val forkjoinPool = new ForkJoinPool(
    1, //  simulate compete
    ForkJoinPool.defaultForkJoinWorkerThreadFactory,
    null,
    false) // you cant switch to false, it only impact the sequence of forked tasks

  val spawnTask = new SpawnTask(0)
  val StarvationExternalTask = new FakeTask(911) //!!!

  forkjoinPool.execute(spawnTask)
  forkjoinPool.execute(StarvationExternalTask)

  new Thread("reporter") {
    override def run() {
      while (true) {
//        println( s"""
//             StealCount            -> ${forkjoinPool.getQueuedTaskCount}
//             QueuedTaskCount       -> ${forkjoinPool.getQueuedTaskCount}
//             QueuedSubmissionCount -> ${forkjoinPool.getQueuedSubmissionCount}""")
        Thread.sleep(5000)
      }
    }
  }.start()

}

class SpawnTask(id: Int) extends ForkJoinTask[Unit] {
  override def getRawResult: Unit = ()

  override def setRawResult(value: Unit): Unit = ()

  override def exec(): Boolean = {
    println(s"SpawnTask! ID: $id")
    Thread.sleep(5000)
    new SpawnTask(id + 1).fork()
    new FakeTask(id + 1).fork()
    true
  }
}

class FakeTask(id: Int) extends ForkJoinTask[Unit] {
  override def getRawResult: Unit = ()

  override def setRawResult(value: Unit): Unit = ()

  override def exec(): Boolean = {
    println(s"FakeTask ID: $id")
    true
  }
}

```
I think it does NOT involve different dispatcher!

**Comment by drewhk** [comment]
Hmm, I think @MQ-EL you might be on the right track.

**Comment by drewhk** [comment]
Technically this means that self-messages are not a necessary condition to trigger this problem, probably it is possible to construct other starvation-prone scenarios.

**Comment by drewhk** [comment]
Minimized the test, and it is reproducable very reliably. If I add a simple random tie breaking function that sometimes forces a call to exectue compared to fork, then it passes.

**Comment by MQ-EL** [comment]
Aother starvation scenarios:
```scala
package com.eastsoft.sample.test

import akka.actor.{ActorRef, ActorSystem, Props, Actor}
import com.typesafe.config.ConfigFactory

object ActorHang {

  val config = ConfigFactory.parseString(
    """
      |anotherhang {
      |
      |  task-dispatcher {
      |    mailbox-type = "akka.dispatch.SingleConsumerOnlyUnboundedMailbox"
      |    throughput = 100
      |    fork-join-executor {
      |      parallelism-factor = 2
      |      parallelism-max = 1
      |      parallelism-min = 1
      |    }
      |  }
      |}
    """.stripMargin
  )
  
  val system = ActorSystem("hang", config)
  val ping = system.actorOf(Props(new Player("ping")).withDispatcher("anotherhang.task-dispatcher"))
  val pong = system.actorOf(Props(new Player("pong")).withDispatcher("anotherhang.task-dispatcher"))
  
  
  def main(args: Array[String]): Unit = {
    
     ping ! Ball(pong)

    val referee = system.actorOf(Props(new Referee).withDispatcher("anotherhang.task-dispatcher"))

    Console.println("Stop the dispute!")
    referee  ! Stop

    system.awaitTermination()
  }

  case class Ball(target: ActorRef) 
  
  class Player(val name: String) extends Actor {
    override def receive: Receive = {
      case Ball(target) =>
        println(s"$name")
        target ! Ball(self)
    }
  }
  
  case object Stop
  
  class Referee extends Actor {
    override def receive: Receive = {
      case msg @ Stop =>
        println("Received command STOP!")  // Never happens  
    }
  }

}

```
I would like see your test code (If I add a simple random tie breaking function...)

**Comment by drewhk** [comment]
I am running benchmarks to see if we can tweak the current state with some randomization, or if we need to revert.

**Comment by MQ-EL** [comment]
I think forked task  is more effective than execute, because FJ thread doesn't needed scan other queues. Can limit the max number of the succession call of fork method?  Record in thread local variable?

**Comment by MQ-EL** [comment]
By the way, in akka remote system, why not use akka.io instead of netty?I think it better than Netty reactive thread model .

**Comment by drewhk** [comment]
Summary of today. I consider this a critical bug. As I postulated, and @MQ-EL demonstrated, no self-messages are needed to reproduce the problem, only a sufficient amount of busy actors communicating between each other on a fj-dispatcher and an actor on a different dipatcher trying to communicate with them.

I tried to salvage the scenario by adding a "yield" function that replied true in a certain fraction of cases, forcing a task to take the normal external submission path instead of the fork.  I tried a very efficient PRNG (XorShift) and a simple modulo counter, using them both consistently (using an atomic field) and "best-effort" mode using a non-volatile var (we don't care for stale reads or missed writes for yields). All the four trials resulted in very similar benchmark results, and only slightly slower than the 2.3.10 version.

However. When running the starvation reproducer test in multiple iterations, the time of the test evolved like:
 - 2.3.8 version takes time X
 - 2.3.10 version fails
 - yield 1/2 of times takes 2X
 - yield 1/4 of times takes 4X
 - etc.

This means that yielding will still significantly throttle the throughput of inter-dispatcher communication.

Another note is that the benchmark differences between 2.3.8 and 2.3.10 were much less for me than Viktor measured.

Overall, I think the current situation is rather bad, and simple yielding strategy does not give a satisfying solution. I strongly recommend reverting the current optimization.

**Comment by drewhk** [comment]
@MQ-EL 

> By the way, in akka remote system, why not use akka.io instead of netty?I think it better than Netty reactive thread model .

We are planning to eventually implement akka-remoting on streams. We need to finish some other things first, though.

**Comment by drewhk** [comment]
If everyone agrees, I will prepare a PR that reverts to 2.3.8 implementation and add reproducer tests for the starvation scenarios so we don't make the same mistakes in the future.

**Comment by drewhk** [comment]
Also thanks @clockfly @MQ-EL for the detective work, you saved us many hours of debugging.

**Comment by MQ-EL** [comment]
It's my pleasure!

**Comment by hepin1989** [comment]
is there any way to setup the inner and outer execution ratio?

**Comment by hepin1989** [comment]
the code in java8 changed.
```
    /**
     * Top-level runloop for workers, called by ForkJoinWorkerThread.run.
     */
    final void runWorker(WorkQueue w) {
        w.growArray();                   // allocate queue
        int seed = w.hint;               // initially holds randomization hint
        int r = (seed == 0) ? 1 : seed;  // avoid 0 for xorShift
        for (ForkJoinTask<?> t;;) {
            if ((t = scan(w, r)) != null)
                w.runTask(t);
            else if (!awaitWork(w, r))
                break;
            r ^= r << 13; r ^= r >>> 17; r ^= r << 5; // xorshift
        }
    }
```

**Comment by drewhk** [comment]
LOL, it also uses xorshift. With the exact same parameters as I did. But yes, that could be an option. But I would do that for 2.4 and avoid risking instability in 2.3 and reverting there.

**Comment by hepin1989** [comment]
LOL,cause you are so clever.I know in netty,we could specify the read/write ratio.I don't know could we do that in FJ pool.

**Comment by MQ-EL** [comment]
In my PC, i test jdk1.7.0_75, 1.8.0_25, 1.8.0_45 (as @hepin1989 showed code version) 

All the result is the same :  **STARVATION**

@hepin1989, you can look at implemention of `w.runTask`.

 I think it is just to tuning steal strategy for improve efficiency(like PRNG, etc). I think this unfairness is deliberate in FJ

**Comment by hepin1989** [comment]
Did you change the the executor to a custom one which make use of the jdk's but not the scala's.

**Comment by MQ-EL** [comment]
Just simple, change the import clause:

```scala
//import scala.concurrent.forkjoin.{ForkJoinPool, ForkJoinTask}
import java.util.concurrent.{ForkJoinPool, ForkJoinTask}
```

**Comment by drewhk** [comment]
The optimization has been reverted.

**Comment by clockfly** [comment]
Good to hear this is solved in new version.

:+1:

**Comment by hepin1989** [comment]
@clockfly will your gearpump based on akka-stream?I know that mogujie is using akka too.
